### DISCUSSION TITLE
How to set current namespace for the CLI?

### DISCUSSION THREAD

**Discussion by eisig** [fluxcd/flux2#1284] [Q&A] [answered]
We would like to install flux resouce to different namespace for different team. 
How to set current namespace, so  we do not have to pass '-n ' everytime? (like kubectl)

**Comment by kingdonb** [comment]
This is a good question. In `fluxctl` there was `--k8s-fwd-ns` which could be set as an environment variable `FLUX_FORWARD_NAMESPACE`.

Now, I'm not sure there is a way to do this -- you can use `alias flux='flux -n my-tenant-ns'` – does that help?

**Reply by stefanprodan** [reply]
The `fluxctl --k8s-fwd-ns` is not the same thing as `flux --namespace`, one sets the namespace where the Flux pod is, the other one sets the namespace where the CRs are. There is no need for `--k8s-fwd-ns` in v2 since the HTTP API is gone.

**Reply by kingdonb** [reply]
The `flux` binary does not respect the current kubeconfig context for default namespace setting, it just uses `flux-system`. This is inconvenient, since there is no way to set a context namespace for flux.

If you are using `kubens`, or a similar workflow with `kubectl` which is like this:

```
kubectl config set-context --current --namespace=where-flux-resources-are
```

so that you can run `flux get kustomizations` and see the kustomizations in that namespace.

This is what I think is being requested. Personally I like that `flux get kustomizations` always returns the resources in `flux-system` how it works today, but if you are in a multi-tenant cluster and don't have access to `flux-system` this behavior is suddenly not so convenient.

**Reply by stefanprodan** [reply]
Like kubectl, flux CLI could have a config file and offer commands to generate and update the config. I'm not for adding such complexity to the CLI while it's not GA.

**Reply by kingdonb** [reply]
It could for instance make sense if each repo could have its own `.fluxconfig` declaring the namespace of the associated resources there. I agree there is a lot of complexity to explore in how flux can be configured to work in different contexts. You could have a fluxconfig in your home directory, and another one per-project. This can get very complicated very quickly.

This is also a low value feature, nice to have feature since `alias flux="flux -n my-tenant"` provides everything you need for at least 80% of users, who will most likely only ever work with one tenant on an average day, while the other 20% is likely to go on needing to use `flux -n namespace` for every day unpredictable needs that might not be as simple as one config per repository.

It is still awkward for tenant users currently, (but I honestly don't have a good proposal for how to make it better.)

**Reply by hiddeco** [reply]
There is no relationship between `flux` and any repository, making this in my opinion a cumbersome addition as one would have to `cd` into the locally cloned repository first.

Instead, it would in my opinion be better to have a "fluxconfig" configuration file type that looks more like a "kubeconfig", including support for contexts, stored on the user's local machine (in `XDG_CONFIG_HOME` or MacOS / Windows equivalent) .

This way, a user could for example create a new context with a mapping to a kubeconfig and the primary `GitRepository` source on this cluster, opening the space for `flux create` commands that do a Git write instead of an apply as well.

**Comment by stefanprodan** [comment, marked as answer]
Changing the default namespace from `flux-system` to some other value using the `FLUX_SYSTEM_NAMESPACE` environment variable has been implemented in https://github.com/fluxcd/flux2/pull/2034
